### ISSUE TITLE
Error: #3005 (activateOnceError): You need to be activate Adapty once

### ISSUE THREAD

**Issue by WrathChaos** [adaptyteam/AdaptySDK-React-Native#42]
Hello,
We just upgraded to version `2.2.1`. We migrated to every function and new stuff. 

```tsx

import { adapty } from "react-native-adapty";

useEffect(() => {
  (async () => {
    await adapty.activate(ADAPTY_SDK_PUBLIC_KEY);
  })();
}, []);

```

Here is my activation code on `App.tsx`. 

Each time it gives me the `Error: #3005 (activateOnceError): You need to be activate Adapty once` and we do not call `adapty.activate` else where for sure. 

Even sometimes it gives `#2002 (notActivated): The Adapty is not activated` but probably it is because of the development environment and stuff.

What should I do to fix this warning? It was all working great on the `1.3.13` version. I’m thinking to downgrade the version.

**Comment by divanc** [comment]
Hey!
I do not see any way to get `notActivated` after `activateOnceError`. I'm going to need a reproducible snippet for that, if you are able to make one.

About `activateOnceError`. This is not an incorrect behavior as is. You activate adapty **once you cold boot your application**. Hot reload/Fast Refresh technology [re-executes useEffect and most other hooks on every refresh](https://reactnative.dev/docs/fast-refresh#fast-refresh-and-hooks), whereas activation status persists.

Now, I'm to blame for not giving enough thought about this being a problem. I persisted only one call with `useRef<boolean>` or just wrapped `activate()` into a `try {} catch {}` block to ignore the error.

I'll try to make something smart for the following release, that would not incorporate any complex logic. In the meantime, can you try any of my proposed solutions?

**Comment by WrathChaos** [comment]
Thank you for this awesome comment and suggestions :) 

I will try to ignore it and also try to reproducible example for you

**Comment by WrathChaos** [comment]
Hey again @divanc 
https://github.com/WrathChaos/adapty-v2-error-example

Here is the reproducible example. All you need to do is replace the api-key and run the ios. 

The try-catch console.log will appear each time. 

I put my `useIAP` custom hook just in-case you want to check it out if something wrong on my code. But it always happens even without my custom hook either. 

Can you please take a look at it? 

Maybe is it because of my react-native version? (0.68.5)

**Comment by WrathChaos** [comment]
Since I found you, let me ask an off-topic about adapty :) 

I wonder what is the best practice for checking and setting the user's premium data. Do I need to check and save the user's premium profile each time on my backend side? Or should I even need to save this data for each month? Like saving the receipt and stuff?

**Comment by WrathChaos** [comment]
Thank you so much! @divanc :)

**Comment by divanc** [comment]
> Since I found you, let me ask an off-topic about adapty :)
> 
> I wonder what is the best practice for checking and setting the user's premium data. Do I need to check and save the user's premium profile each time on my backend side? Or should I even need to save this data for each month? Like saving the receipt and stuff?

I've asked my colleagues to provide you a better answer, than what I can give you 😅
Here is what I think:

If you have fully integrated Adapty in your paid content chain, then there is no need to duplicate anything to your side for validating purposes. At least, I haven't developed an app at such a scale. Adapty works directly with native StoreKits, so if a StoreKit and Adapty API are available, then the chances that you would get an incorrect status is minimal. If Adapty API is unavailable (I cannot remember such an occur in the last 2 years), then SDK would provide you a cached result. 

This is another nice point. Adapty SDK does caching and a number of other things under the hood. If you would wait for your API to approve subscription status, then what about the case when user does not have internet? Adapty handles such cases to the best of my knowledge.

The only question is *why* really. If you can answer that, then there might be a reason. You just need to know that it would take quite some time to implement a nice validation solution and then debug it. 

I may have misunderstood your question: please, correct me if I have given you an unrelated reply.

**Comment by divanc** [comment]
Testing your reproducible...

**Comment by divanc** [comment]
I did the following:

1. `yarn install`
2. `cd ios && pod install`
3. added my token
4. `xed -b ios` (launches Xcode)
5. Added my Bundle ID
6. Added my team for signing cert
7. Removed previous watchman logs
8. `yarn start`
9. Run on device in Xcode

I get the following:
<img width="1338" alt="image" src="https://user-images.githubusercontent.com/16625670/214681103-5681aac3-05dc-4027-9567-01393b400924.png">


It is rather strange, that it somehow without Swift bridging header...

**Comment by divanc** [comment]
'Server error' is about not finding paywall, I guess?

**Comment by divanc** [comment]
Localization is not good in that manner

I totally agree that activate is throwing an error more often, than preferable. Going to do something with it

**Comment by divanc** [comment]
While I was unable to get `notActivated` after `activateOnceError`, you can try an experimental feature in v2.2.1 — `lockMethodsUntilReady`. It makes all your calls wait until `activate` is resolved, so there wouldn't be any race condition.
```js
adapty.activate('KEY', {
  lockMethodsUntilReady: true,
});
```

**Comment by WrathChaos** [comment]
@divanc 
I do not get Server error, actually I am able to fetch my paywall, products and even the profile has a premium or not. 

I tried to use the `lockMethodsUntilReady` one but it just throws much more error than activateOnceError. So it is not a solution for me. I could not even call the paywall with this prop.

**Comment by divanc** [comment]
Testing with my paywall IDs...

**Comment by WrathChaos** [comment]
Do you think anything else which I can provide? @divanc

**Comment by divanc** [comment]
Hey, sorry for being absent! Pleases check what I've found with your app:

Let's say, there are two purchase blocks in your app:
1. activating Adapty (my version)
```tsx
useEffect(() => {
    async function fetch() {
      try {
        console.log('Registering adapty');
        await adapty.activate('TOKEN');
        console.log('Register: SUCCESS');
      } catch (error) {
        if (error instance AdaptyError) {
          console.log('Error activating', error.message);
        }
      }
    }
    fetch();
  }, []);
```

2. fetching a paywall, initiating your sub-SDK
```tsx
const {isPremiumActive} = useIAP();
```

And I had somewhat consistent behavior: if I have declared activation promise before paywalls promise, then error is not raised. So in your `App.tsx` I have **1 → 2**, Doing **2 → 1** seemed to result in this error you face.

Now I certainly understand that this is an insecure way of activating an sdk, but this somehow works until I can make something better! Also in your app `lockMethodsUntilReady` works just nicely for me. Can you provide some logs on that?

### 2 → 1 logs
```log
JS Registered
Running "adapty" with {"rootTag":1,"initialProps":{}}
Fetching Paywalls...
Registering adapty
Paywall error:  Error: #2002 (notActivated): The Adapty is not activated
Register: SUCCESS
```

### 1 → 2 logs
```log
JS Registered
Running "adapty" with {"rootTag":1,"initialProps":{}}
Registering adapty
Fetching Paywalls...
Register: SUCCESS
Paywalls: SUCCESS
Fetching products...
Products: SUCCESS
```

**Comment by divanc** [comment]
I plan to move activation outside of components to be hooked by JS interpreter before components mount, like on picture below:
<img width="490" alt="image" src="https://user-images.githubusercontent.com/16625670/215318427-f19e6b19-b0b1-4be0-ab96-0abd58ba1b57.png">

But, I still need
* To read a lot to understand how it works deeper, so background caching, event emitter and stuff wouldn't break
* Probably to introduce another activation design in that case, as I consider this of another design pattern

**Comment by WrathChaos** [comment]
Hello @divanc 
Thank you so much for the detailed answer! 

I tried both ways to activate the the adapty but I still got the errors something like this: 
![CleanShot 2023-01-29 at 19 53 27](https://user-images.githubusercontent.com/6813862/215342048-1df97513-9a3b-4820-8883-8d346989dda4.png)


I simply copy-pasted this into my App.tsx

useEffect(() => {
    async function fetch() {
      try {
        console.log('Registering adapty');
        await adapty.activate('TOKEN');
        console.log('Register: SUCCESS');
      } catch (error) {
        if (error instance AdaptyError) {
          console.log('Error activating', error.message);
        }
      }
    }
    fetch();
  }, []);

```

and then I have a straightforward react-navigation that just navigates my HomeScreen and I am simply calling my `useIAP` custom hook. 

```
  const { isPremiumActive } = useIAP();
```

In this hook is almost identical with the given repo example. 

I got it, it is something wrong with my implementation, but I don't know how to wait for the activation because it simply does not wait for the activation and just tried to do something with the adapty. 

I also tried to move it outside of the components to be hooked by the JS interpreter before the components mount on App.tsx but simply it gives the same error. 


But this way, it looks like there is no way to wait for the adapty activation to re-use the adapty inside the other components, screens?

**Comment by divanc** [comment]
Let's reiterate — we're facing two problems:
1) 3005 (already activated) throws more often than it should (in your example **should not** throw at all)
2) 2002 (not activated) is being thrown in your example in paywalls section, although it should not. lockUntilReady thing does not help

Are there two all we discuss so far?

**Comment by divanc** [comment]
> Hello @divanc Thank you so much for the detailed answer!
> 
> I tried both ways to activate the the adapty but I still got the errors something like this: ![CleanShot 2023-01-29 at 19 53 27](https://user-images.githubusercontent.com/6813862/215342048-1df97513-9a3b-4820-8883-8d346989dda4.png)
> 
> I simply copy-pasted this into my App.tsx
> 
> ```tsx
> useEffect(() => {
>     async function fetch() {
>       try {
>         console.log('Registering adapty');
>         await adapty.activate('TOKEN');
>         console.log('Register: SUCCESS');
>       } catch (error) {
>         if (error instance AdaptyError) {
>           console.log('Error activating', error.message);
>         }
>       }
>     }
>     fetch();
>   }, []);
> ```
> 
> and then I have a straightforward react-navigation that just navigates my HomeScreen and I am simply calling my `useIAP` custom hook.
> 
> ```tsx
>   const { isPremiumActive } = useIAP();
> ```
> 
> This hook is almost identical with the given repo example.
> 
> However, when I call the adapty activation into component's useEffect like a componentDidMount, I just get this error:
> 
> ![CleanShot 2023-01-29 at 20 05 07](https://user-images.githubusercontent.com/6813862/215343173-c31e2df7-7a2e-4441-acbb-3cf15b91ff3c.png)
> 
> Outside the component is making lots of unhandled promises :)
> 
> I got it, it is something wrong with my implementation, but I don't know how to wait for the activation because it simply does not wait for the activation and just tried to do something with the adapty.
> 
> I also tried to move it outside of the components to be hooked by the JS interpreter before the components mount on App.tsx but simply it gives the same error.
> 
> But this way, it looks like there is no way to wait for the adapty activation to re-use the adapty inside the other components, screens?
> 
> I can use it with try-catch to ignore the error but I think it should not be like this right? :)

In your attached screenshot I can see an interesting thing: you have an `unhandled promise rejection` about activating more than once. Also you said, that you've taken my `fetch` function to your flow. But my `fetch` function should not be able to throw an unhandled error, since it wrap all the body in a try-catch block. 
Can you check, there is no other place, where you try to activate?

**Comment by divanc** [comment]
@WrathChaos please, check out v2.3.0!

With this release I was able to solve both of your errors on my end.
To prevent paywall fetching before activation mutex, I moved it outside of component as you can see in [example App.tsx](https://github.com/adaptyteam/AdaptySDK-React-Native/blob/0e5f20e83c9595291230dbc0df222bee142dba02/examples/SubscriptionsJS/App.js#L26).

You can call this activation in any JS file, just make sure to import it to your core.

**Comment by WrathChaos** [comment]
Wow! Thank you so much @divanc 
I will try it out today!

**Comment by WrathChaos** [comment]
More information:

When I use `lockMethodsUntilReady` more unhandled promise rejection are appears. 

![CleanShot 2023-02-02 at 20 59 58](https://user-images.githubusercontent.com/6813862/216405060-e79376d6-b172-48ce-a3e8-6896a0b8777c.png)

If I do not use `lockMethodsUntilReady`, try-catch can handle the error and I am still able to fetch profile, paywall etc

![CleanShot 2023-02-02 at 21 01 58](https://user-images.githubusercontent.com/6813862/216405875-6b7d6be8-4330-4499-b2a8-9a43b15fd7cf.png)

**Comment by divanc** [comment]
Sorry, had to jump on another task. Diving in...

**Comment by WrathChaos** [comment]
@divanc Is there any news? :)

**Comment by divanc** [comment]
Sorry for not responding sooner, had to be offline for several days 😓
Your example made my log go crazy for a bit too. Although I tweaked your code a little and the problem was solved for me:

1. Instead of `'TOKEN'` string I placed my test token, same for paywallID, Apple Team ID and bundle ID (duh)
2. I removed second activation of adapty in `useEffect`, that seemed to cause that hustle in logs. Here is my diff:

<img width="2056" alt="image" src="https://user-images.githubusercontent.com/16625670/219943084-3217a859-b30e-4c84-a0a0-9bbad4ab7086.png">

In your repo, this is /App.tsx:46

**Comment by divanc** [comment]
Can you confirm that this works for you?

**Comment by divanc** [comment]
@WrathChaos ping for the info 🏓

**Comment by WrathChaos** [comment]
Hey @divanc 
I will check it out this week :) sorry for the late response, some health issues

**Comment by divanc** [comment]
Hang in there! 💊

**Comment by divanc** [comment]
v2.3.5 MUST solve your issue! I'm going to write some more here today

**Comment by WrathChaos** [comment]
I will try it out today and let you know

**Comment by WrathChaos** [comment]
@divanc 
It works! Thank you so much for helping sir!

🥇

**Comment by divanc** [comment]
If you wonder: 
The memoization was a nice pattern to use, but it turned out to be inconsistent in my realization. I've made argument memoization on JS side, which totally worked on my end. Hot reload haven't dropped the variables values, BUT for some bundler occasionally decided to drop JS variables and perform full reload cycle.
So on a native side Adapty was active, whereas on JS side it was considered never called. In v2.3.6 I've moved memoization to native layers, duplicating the code. But hey, this is the only solution, where RN bundler would certainly not interfere
